Mathlib/GroupTheory/MonoidLocalization/Basic.lean

```diff
@@ -990,6 +1033,30 @@ theorem map_map {A : Type*} [CommMonoid A] {U : Submonoid A} {R} [CommMonoid R]
   rw [← f.map_comp_map (k := k) hy j hl]
   simp only [MonoidHom.coe_comp, comp_apply]
 
+@[to_additive] theorem map_injective_of_surjOn_or_injective
+    (or : (S : Set M).SurjOn g T ∨ Injective k) (hg : Injective g) :
+    Injective (f.map hy k) := fun z w hizw ↦ by
+  set i := f.map hy k
+  have ifkg (a : M) : i (f a) = k (g a) := f.map_eq hy a
+  let ⟨z', w', x, hxz, hxw⟩ := surj₂ f z w
+  have : k (g z') = k (g w') := by rw [← ifkg, ← ifkg, ← hxz, ← hxw, map_mul, map_mul, hizw]
+  obtain surj | inj := or
+  · have ⟨⟨c, hc'⟩, eq⟩ := k.exists_of_eq this
+    obtain ⟨c, hc, rfl⟩ := surj hc'
```

Comment:
Why do you use `let`, `have`, `obtain` in four different lines for the "same" tactic?

Reply:
The `let` should be `have`, thanks. I only use `obtain` where `have` doesn't work.